seq_tools/validation/c620_rg_id_in_bam_check.py

```diff
@@ -0,0 +1,108 @@
+        super().__init__(ctx, metadata, __name__)
+
+    @BaseChecker._catch_exception
+    def check(self):
+        # get all RG ID from BAM(s)
+        files_in_metadata = self.metadata['files']
+
+        rg_id_in_bams = {}
+        for f in files_in_metadata:
+            f = f['fileName']
+            if not f.endswith('.bam'):  # not a BAM, skip
+                continue
+
+            bam_file = os.path.join(self.submission_directory, f)
+
+            # retrieve the @RG from BAM header
+            header = subprocess.check_output(
+                ['samtools', 'view', '-H', bam_file],
+                stderr=subprocess.STDOUT
+            )
+            header_array = header.decode('utf-8').rstrip().split('\n')
+
+            for line in header_array:
+                if not line.startswith("@RG"):
+                    continue
+                rg_array = line.rstrip().replace('\t', '\\t')
+
+                # get rg_id from BAM header
+                rg_id_in_bam = ':'.join([
+                    kv for kv in rg_array.split('\\t') if kv.startswith('ID:')
+                ][0].split(':')[1:])
+
+                if f not in rg_id_in_bams:
+                    rg_id_in_bams[f] = []
+
+                # don't need to check rg_id uniquessness as it's checked already elsewehere earlier
+                rg_id_in_bams[f].append(rg_id_in_bam)
+
+        read_groups = self.metadata['read_groups']
+
+        offending_ids = {}
+        rg_id_in_bam = []
+        for rg in read_groups:
+            filename = rg['file_r1']
+            if not filename.endswith('.bam'):
+                continue  # skip if not BAM
+
+            if rg.get('read_group_id_in_bam', None) is not None:
+                rg_id_in_bam.append(rg['read_group_id_in_bam'])
+                continue
+
+            if rg.get('submitter_read_group_id') not in rg_id_in_bams.get(filename, []):
+                if filename not in offending_ids:
+                    offending_ids[filename] = []
+                offending_ids[filename].append(rg.get('submitter_read_group_id'))
+        
+        if offending_ids:
+            msg = []
+            for k, v in offending_ids.items():
+                msg.append("BAM %s: %s" % (k, ', '.join(sorted(v))))
+
+            self.status = 'INVALID'
+            message = "'read_group_id_in_bam' are NOT provided in metadata and " \
+                "'submitter_read_group_id' specified in 'read_groups' section of the metadata are not found " \
+                "in BAM file. Offending ID(s): %s" % '; '.join(msg)
+            self.message = message
+            self.logger.info(f'[{self.checker}] {message}')
+
+        elif rg_id_in_bam:
+            self.status = 'VALID'
+            message = "'read_group_id_in_bam' are provided in metadata and the current checker is skipped. Please refer to the next checker for further information."
+            self.message = message
+            self.logger.info(f'[{self.checker}] {message}')
+
```

Comment:
don't need this block

Reply:
I feel it is not good to send the message `'submitter_read_group_id' in metadata matches RG ID in BAM check: VALID` as we do NOT check whether they match or not.

Reply:
good point, but yes and no. Even when `elif rg_id_in_bam` is true, it does not mean `submitter_read_group_id` check has been skipped. It's possible when there are multiple read groups, some provided `read_group_id_in_bam` some did not.

A simple way to improve clarity here is to make the message more clear: `For a read group, when 'read_group_id_in_bam' is not provided, 'submitter_read_group_id' in metadata matches RG ID in BAM check: VALID`

This actually brings up a general question I started to realize when started to write validation checks for BAM. For a pure FASTQ submission, these checks are not applicable. So reporting the check result as `VALID` could be misleading. Maybe we should change the status `VALID` to `PASS` for all checks? Also, we can add BAM file name(s) in the message to be explicit about which BAMs have been validated.

Reply:
yeah, I agree. I think `PASS` is better than `VALID`